### PR TITLE
First commit

### PR DESCRIPTION
**Pull request by wzshiming**
fake-kubelet-manager simulates the behavior of multiple kubelets

## Feature

The fake-kubelet-manager only keeps the heartbeat of the Node and fills in the status of the Node and Pod when necessary

### Compare [wzshiming/fake-kubelet](https://github.com/wzshiming/fake-kubelet) Will No Create Node 
The `kwok-controller` will no longer be responsible for creating Nodes, and will only handle Nodes created by others.

### CODE REVIEW

File: test/fake-kubelet-manager/fake-kubelet-manager.test.sh [hunk 1]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env bash
```
Comment (wzshiming) [inline test/fake-kubelet-manager/fake-kubelet-manager.test.sh:1]: This script integrates and tests the full functionality of this commit

File: cmd/fake-kubelet-manager/main.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+	err = ctr.Start(ctx)
+	if err != nil {
+		logger.Fatalln(err)
+	}
+
+	<-ctx.Done()
+}
+
+func Server(ctx context.Context, address string) {
```
Comment (Huang-Wei) [inline cmd/fake-kubelet-manager/main.go:140]: `Serve` may read better?

File: kustomize/fake-kubelet-manager/fake-kubelet-manager-clusterrole.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (Huang-Wei) [inline kustomize/fake-kubelet-manager/fake-kubelet-manager-clusterrole.yaml:1]: rename this file to `kustomize/kwok/clusterrole.yaml` (applies elsewhere)

**Review by Huang-Wei — COMMENTED** [review]
First round of review.

File: cmd/fake-kubelet-manager/main.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
+/*
```
[the remark this answers is not in the document]
Comment (wzshiming) [reply cmd/fake-kubelet-manager/main.go:1]: `kwok/kowk` this name I was planning to reserve for https://github.com/wzshiming/fake-k8s,
then use `kwok create cluster` like kind

The name is `kwok/fake-kubelet-manager` is not really appropriate or replaced with `kwok/kwok-controller`?

File: images/fake-kubelet-manager/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM docker.io/library/golang:1.18-alpine AS builder
+WORKDIR /go/src/sigs.k8s.io/kwok
+COPY . .
+
+# Check that if no vendor exists, install git in preparation for pulling the vendor later
+RUN ls vendor || apk add -U git
```
[the remark this answers is not in the document]
Comment (wzshiming) [reply images/fake-kubelet-manager/Dockerfile]: I haven't pushed the `vendor` directory in the current project, for a better review.
Worried about not having git errors in the ci build, maybe if I switch to a base image with git, then I won't have that problem.

File: cmd/fake-kubelet-manager/main.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	ctx    = context.Background()
+	logger = log.New(os.Stderr, "[kwok/fake-kubelet-manager] ", log.LstdFlags)
+)
+
+func init() {
+	pflag.StringVar(&cidr, "cidr", cidr, "CIDR of the pod ip")
+	pflag.IPVar(&nodeIP, "node-ip", nodeIP, "IP of the node")
+	pflag.BoolVar(&allNodeTakeOver, "all-node-take-over", allNodeTakeOver, "Take over all nodes, there should be no nodes maintained by real Kubelet in the cluster")
+	pflag.StringVar(&nodeTakeOverAnnotationSelector, "node-take-over-annotation-selector", nodeTakeOverAnnotationSelector, "Selector of nodes that with this annotation will to take over")
+	pflag.StringVar(&statusCustomAnnotationSelector, "status-custom-annotation-selector", statusCustomAnnotationSelector, "Selector of pods that with this annotation will no longer maintain status and will be left to others to modify it")
```
[the remark this answers is not in the document]
Comment (wzshiming) [reply cmd/fake-kubelet-manager/main.go]: - they don't want to change the existing API objects' labels/annotations. In this case, maybe we should provide annotation AND label selector?
        - they're fine with applying extra labels/annotation. In this case, they just apply with default labels/annotation, and leave this parameter as nil.

I also added a set of flags for the label
--all-node-manage
--node-manage-annotation-selector
--node-manage-label-selector
--status-custom-annotation-selector
--status-custom-label-selector

        - plain CLI argument may not express the semantics well (moreover, how to express a semantics like "foo!=bar" or "foo not in [bar, baz]"

https://github.com/kubernetes/apimachinery/blob/899984fb2d224c4a86ca7906e38754a3a62f8c41/pkg/labels/selector.go#L846-L884

CLI should be able to express semantics well, just like kubectl.


    With all the above said, I think we may follow up with an enhancement to standardize the parameters (like scheduler ComponentConfig), like

Do we need to define a configuration for kwok? Is it a local file, or configmap or CRD?

File: cmd/fake-kubelet-manager/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply cmd/fake-kubelet-manager/main.go]: Hmm... I think `kwok` (the controller) is the core and essential part of this repo, so I'd like to reserve kwok for it.

Regarding other supplimental CLI tools, `kwokctl` may read better?

File: cmd/fake-kubelet-manager/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply cmd/fake-kubelet-manager/main.go]: > Do we need to define a configuration for kwok?

I'd like to have a version configuration to organize the parameters. It's a local file, not configMap/CRD, just like [scheduler config](https://kubernetes.io/docs/reference/scheduling/config/).

But defining CLI args is a good start.

File: pkg/kwok-controller/controller/controller.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
+/*
```
Comment (Huang-Wei) [inline pkg/kwok-controller/controller/controller.go:1]: The file name can be `pkg/controllers/controller.go`.

File: pkg/kwok-controller/controller/node_controller.go [hunk 8]
```diff
@@ -0,0 +1,401 @@
+/*
```
Comment (Huang-Wei) [inline pkg/kwok-controller/controller/node_controller.go:1]: The file name can be `pkg/controllers/node_controller.go`.

File: cmd/fake-kubelet-manager/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (wzshiming) [reply cmd/fake-kubelet-manager/main.go]: #4

I created a new issue, to discuss the next step in defining the configuration

File: cmd/kwok-controller/main.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+	// If a Pod is on a managed Node and has this annotation status will not be modified
+	statusCustomAnnotationSelector = ""
+	// If a Node being managed has this label it will only keep its heartbeat not modify another status
+	// If a Pod is on a managed Node and has this label status will not be modified
+	statusCustomLabelSelector = ""
+
+	serverAddress = ""
+	master        = ""
+	kubeconfig    = getEnv("KUBECONFIG", "")
+	logger        = log.New(os.Stderr, "[kwok/kwok-controller] ", log.LstdFlags)
```
[the remark this answers is not in the document]
Comment (wzshiming) [reply cmd/kwok-controller/main.go:62]: #2
I created a new issue, to evaluate the other promising contextual logger.

File: cmd/kwok-controller/main.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+)
+
+func init() {
+	pflag.StringVar(&cidr, "cidr", cidr, "CIDR of the pod ip")
+	pflag.IPVar(&nodeIP, "node-ip", nodeIP, "IP of the node")
+	pflag.BoolVar(&allNodeManage, "all-node-manage", allNodeManage, "Manage all nodes, there should be no nodes maintained by real Kubelet in the cluster")
```
[the remark this answers is not in the document]
Comment (wzshiming) [reply cmd/kwok-controller/main.go]: `disregard-pods-with-annotation-selector`
This works for both Pods and Node, and it makes kwok no longer manage its Status

maybe it should be

- `disregard-status-with-annotation-selector` All node/pod status excluding the ones that match the annotation selector will be watched and managed.
- `disregard-status-with-label-selector` All node/pod status excluding the ones that match the label selector will be watched and managed

or

- `disregard-pod-status-with-annotation-selector`
- `disregard-pod-status-with-label-selector`
- `disregard-node-status-with-annotation-selector`
- `disregard-node-status-with-label-selector`

File: cmd/fake-kubelet-manager/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (wzshiming) [reply cmd/fake-kubelet-manager/main.go]: `cmd/controller` -> `registry.k8s.io/kwok/kwok`
`cmd/kwokctl` -> `registry.k8s.io/kwok/kwokctl`
[all-in-one image](https://github.com/wzshiming/fake-k8s/blob/master/images/cluster/Dockerfile) -> `registry.k8s.io/kwok/cluster`

Like this?

**Review by Huang-Wei — COMMENTED** [review]
Good to go to start this project. 🚢 

/lgtm

**Pull request merged**